### PR TITLE
More `null` protection

### LINKED ISSUE #3068: [Bug]Null Exception
**Describe the bug**
Null Exception.
I am trying to generate a version on the branch. It works on the master.

## Expected Behavior
Version must be generated.

## Actual Behavior
Crashes with an exception

## Your Environment

- Version Used: gitversion-osx-arm64-5.9.1-beta1-43
- Operating System and version (Windows 10, Ubuntu 18.04): Mac/Linux
- Link to your project:
- Link to your CI build (if appropriate):
```
  ERROR [04/04/22 21:18:52:16] An unexpected error occurred:
System.NullReferenceException: Object reference not set to an instance of an object.
   at GitVersion.RepositoryStore.<>c.<GetMergeCommitsForBranch>b__37_3(BranchCommit b) in /Users/runner/work/GitVersion/GitVersion/src/GitVersion.Core/Core/RepositoryStore.cs:line 504
   at System.Linq.EnumerableSorter`2.ComputeKeys(TElement[] elements, Int32 count)
   at System.Linq.EnumerableSorter`1.ComputeMap(TElement[] elements, Int32 count)
   at System.Linq.EnumerableSorter`1.Sort(TElement[] elements, Int32 count)
   at System.Linq.OrderedEnumerable`1.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at GitVersion.RepositoryStore.GetMergeCommitsForBranch(IBranch branch, Config configuration, IEnumerable`1 excludedBranches) in /Users/runner/work/GitVersion/GitVersion/src/GitVersion.Core/Core/RepositoryStore.cs:line 483
   at GitVersion.RepositoryStore.FindCommitBranchWasBranchedFrom(IBranch branch, Config configuration, IBranch[] excludedBranches) in /Users/runner/work/GitVersion/GitVersion/src/GitVersion.Core/Core/RepositoryStore.cs:line 321
   at GitVersion.Configuration.BranchConfigurationCalculator.InheritBranchConfiguration(Int32 recursions, IBranch targetBranch, BranchConfig branchConfiguration, ICommit currentCommit, Config configuration, IList`1 excludedInheritBranches)
   at GitVersion.Configuration.BranchConfigurationCalculator.GetBranchConfigurationInternal(Int32 recursions, IBranch targetBranch, ICommit currentCommit, Config configuration, IList`1 excludedInheritBranches) in /Users/runner/work/GitVersion/GitVersion/src/GitVersion.Core/Configuration/BranchConfigurationCalculator.cs:line 54
   at GitVersion.Configuration.BranchConfigurationCalculator.GetBranchConfiguration(IBranch targetBranch, ICommit currentCommit, Config configuration, IList`1 excludedInheritBranches) in /Users/runner/work/GitVersion/GitVersion/src/GitVersion.Core/Configuration/BranchConfigurationCalculator.cs:line 28
   at GitVersion.GitVersionContextFactory.Create(GitVersionOptions gitVersionOptions) in /Users/runner/work/GitVersion/GitVersion/src/GitVersion.Core/Core/GitVersionContextFactory.cs:line 38
   at GitVersion.GitVersionCoreModule.<>c__DisplayClass0_0.<RegisterTypes>b__1() in /Users/runner/work/GitVersion/GitVersion/src/GitVersion.Core/GitVersionCoreModule.cs:line 37
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.get_Value()
   at GitVersion.VersionCalculation.NextVersionCalculator.get_context() in /Users/runner/work/GitVersion/GitVersion/src/GitVersion.Core/VersionCalculation/NextVersionCalculator.cs:line 15
   at GitVersion.VersionCalculation.NextVersionCalculator.FindVersion() in /Users/runner/work/GitVersion/GitVersion/src/GitVersion.Core/VersionCalculation/NextVersionCalculator.cs:line 30
   at GitVersion.GitVersionCalculateTool.CalculateVersionVariables() in /Users/runner/work/GitVersion/GitVersion/src/GitVersion.Core/Core/GitVersionCalculateTool.cs:line 52
   at GitVersion.GitVersionExecutor.RunGitVersionTool(GitVersionOptions gitVersionOptions) in /Users/runner/work/GitVersion/GitVersion/src/GitVersion.App/GitVersionExecutor.cs:line 66
  INFO [04/04/22 21:18:52:16] Attempting to show the current git graph (please include in issue): 
  INFO [04/04/22 21:18:52:16] Showing max of 100 commits
  INFO [04/04/22 21:18:52:26] * 1aeae519 4 hours ago  (origin/ss/ps_enum)
```

@asbjornu Thanks for fixing the previous null exception and this is a new one with the latest build.

### PR DESCRIPTION
**Pull request by asbjornu**
Refactoring of `RepositoryStore` to improve readability, provide better stack traces if an exception occurs, and provide better protection against `null`.

## Related Issue

Does not necessarily fix, but closes #3068 by providing a more informative exception and stack trace, should the same issue reoccur.

## Motivation and Context
Less `NullReferenceException`, and more informative stack traces.

## How Has This Been Tested?
I'm relying on existing tests for coverage.

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by asbjornu** [comment]
@arturcic, perhaps we should disable the tests that require secrets in forks and only run them on `pull_request_target`?

**Comment by arturcic** [comment]
> @arturcic, perhaps we should disable the tests that require secrets in forks and only run them on `pull_request_target`?

can you do that as I'm quite busy with the work

**Comment by asbjornu** [comment]
@voltagex, I hope to get this done by today.

File: src/GitVersion.Core/Core/MainlineBranchFinder.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+internal class MainlineBranchFinder
+{
+    private readonly Config configuration;
+    private readonly ILog log;
+    private readonly IEnumerable<KeyValuePair<string, BranchConfig>>? mainlineBranchConfigs;
+    private readonly IGitRepository repository;
+    private readonly RepositoryStore repositoryStore;
+
+
+    public MainlineBranchFinder(RepositoryStore repositoryStore,
```
Comment (arturcic) [inline src/GitVersion.Core/Core/MainlineBranchFinder.cs:18]: maybe `IRepositoryStore` is better?

**Review by arturcic — APPROVED** [review]

**Comment by voltagex** [comment]
Is it possible to get a release created with this fix so it can be used in Azure Pipelines?

**Comment by arturcic** [comment]
There is some PR @asbjornu is working on, so whenever he says he's done we can release.

**Pull request merged**